### ISSUE TITLE
Uber-issue for Mixin compatibility fails

### ISSUE THREAD

**Issue by luacs1998** [ForgeEssentials/ForgeEssentialsMain#1449]
Replaces #1372.

If you run into any problems, attach full logs.

Known incompatible mods:

- FastCraft
- AppleCore
- Mariculture
- LiteLoader

I need to work on a fix with @mumfrey for this (I shade ASM? Mixin themselves shade ASM? FML PR?)

Please watch SpongePowered/Mixin#39 for updates.

**Comment by luacs1998** [comment]
Received report of LiteLoader not working too. Added to OP.

Seriously, what works?

**Comment by Mumfrey** [comment]
Are **all** the issues related to the ASM split? Because if so then we know there's a simple fix as discussed on the other issue. Also, does setting the MixinTweaker's tweakorder to zero not fix the issue?

**Comment by luacs1998** [comment]
Apologies for the constant disturbance, but from simple triaging of crashlogs received on this issue, all of them seemed to be related to the ASM split.

Setting the TweakOrder to zero did not help either.

**Comment by Mumfrey** [comment]
Well in that case, if we can confirm that having FML itself add the exclusion is a viable fix then I think it's the best way to go. Shading ASM just adds extra unnecessary payload and would cause all the ASM classes to be loaded twice in any case.

I haven't gone down this route yet because I have absolutely zero idea how to get FML set up as a development workspace and haven't had time to find out. I will look into it next week.

**Comment by luacs1998** [comment]
It's okay, I'm already in the process of testing the PR, I'll make it

**Comment by Mumfrey** [comment]
Okay, let me know how it goes.

**Comment by luacs1998** [comment]
Preliminary testing with a monkey-patched Forge jar and latest FE seems to be working fine. I'll get a few more guys to validate the fix and then I'll make the PRs.

**Comment by olee** [comment]
Nice! I will test it in a few minutes, too.

**Comment by ultrasn0w** [comment]
Awesome! 
I fear that a Forge update is required then? The latest Forge branch has some breaking fluid system changes and not every mod has updated to the new system yet...

**Comment by luacs1998** [comment]
Yes, a Forge uodate is required (assuming the PRs are even accepted)

**Comment by olee** [comment]
Confirmed to work with fastcraft

**Comment by ultrasn0w** [comment]
:) Let's see. Sometimes  I needed to move on to the new Forge with my modpack anyway, just delaying it as far as possible because of the many broken mods..
@olee Did you also test with Fastcraft & AppleCore?

**Comment by olee** [comment]
Not yet - will once I find their download sites.

**Comment by olee** [comment]
Tested with FastCraft, AppleCore and Mariculture all present at the same time: No crash :+1:

**Comment by luacs1998** [comment]
https://github.com/luacs1998/FML/commit/91338433fa74e782e237643632de2cc5e17ee280

**Comment by luacs1998** [comment]
PRs merged, now waiting on FML to update on the Forge side...

**Comment by luacs1998** [comment]
PRs merged, now waiting on FML to update on the Forge side...

**Comment by luacs1998** [comment]
Use either Forge 1399 or [this patched Forge 1395](http://bit.ly/1Dr6WWQ) to fix this issue.

**Comment by luacs1998** [comment]
I wish I could close this, butt seeing as we kinda opened a can of worms with the PR (MinecraftForge/FML#655), it'll have to be reopened until they can come up with a workable solution -.-

**Comment by luacs1998** [comment]
A very temporary workaround would be that I shade and package-remap ASM with the current Mixin distributable and FE.

I'll need Abrar's help on this, though.

**Comment by Mumfrey** [comment]
Well I'd like to bundle the remapped ASM with mixin rather than having consumers do it, since the idea would be to centralise the remapped ASM for all mods using Mixin. If it's done at the end of the build chain then all mods using Mixin could end up with their own version of ASM, which isn't really ideal.

**Comment by luacs1998** [comment]
Sure, do what you think is fitting. I just need SpongePowered/Mixin#44 out of the way then I can throw out my own custom tweaker. Looking at some improvements in #1470 anyway.

**Comment by liach** [comment]
Reference #1141, @olee planned 1.8 update for April, obviously now its May. Sadly this had been delayed, but what could we do? Sponge builds are already available, and Sponge plugins came out recently. Your pull request was not reverted in Minecraft Forge master branch, but not in 1.7.10. Making a 1.8 branch could be better for Forge Essentials I think.

**Comment by melekaiah** [comment]
Ok, So mixins is to allow FE to work with sponge and forge at same time?
But it breaks alot of mods in 1.7.10. 

Could we not get a version of FE, that is without the mixin loader support? 
I have not idea what mods there are that require mixin at the moment? is there something cool im missing?

**Comment by liach** [comment]
Nope. Mixin allows Sponge to inject outside forge...
Actually this is a non-issue. You should not use coremods. Coremods could fix it themselves, lexmanos has said that. Or make a new branch/fork?

**Comment by luacs1998** [comment]
Alright, let me clear up some misconceptions.

According to the Mixin about page:

> Mixin is a trait/mixin framework for Java using ASM

What this means is that it provides a lot of helper classes which are useful for the many patches FE makes. Yes, Sponge uses Mixin (and in fact the Mixin project was created because of Sponge). Meaning, it makes my life easier, and improves compatibility with Sponge.

This **DOES NOT** mean FE will be able to run on SpongeAPI. FE is still a forge mod first and foremost, but we need to extend Forge as in quite a few cases Forge does not provide hooks we need. Using Mixin allows our Forge extensions to be compatible with Sponge, of which you will see take shape in 1.8.

However, as seen in this issue, as Mixin requires quite a fair bit of special-casing, it breaks other coremods not coded to work with it. FE is a pioneer in the use of Mixins in a production environemtn (seeing as Sponge just got released last week), so we are encountering issues that would otherwise have been encountered by Sponge when the 1.8 era dawns.

@liach we most probably will not be doing a 1.8 release until we can get 1.7 stable out the door, but we need to get this pesky thorn in the side solved first.

@melekaiah Yes, I understand, it breaks a lot of mods, but I'm working with the stakeholders in getting this issue solved.

**Comment by liach** [comment]
Oh it was really a shame to merge #1314 without any further discussion... Oh hoping mumfrey to look into mixin first. Thanks for your explaination.

P.S. BiomesOPlenty also tried to use mixin but abandoned it at last.

**Comment by luacs1998** [comment]
It was already discussed a long time ago... In fact the mixin port was sitting on my hard drive for 2 months before I made that PR.

Well, such is the price to pay for being a pioneer. Either we do it now, or go to 1.8 and encounter the same unexpected issues there.

Thanks for your concern though.

**Comment by Mumfrey** [comment]
@luacs1998 I pinged you on IRC last night but not sure if you got it. There is a branch [here](https://github.com/SpongePowered/Mixin/tree/feature/shade-asm) for you to try out which shades a renamed ASM into the jar. I have pushed the branched jar (which is now version **0.4**) to my maven repository [here](http://repo.mumfrey.com/content/repositories/mixin/) for your convenience (since I won't be pushing this to master and thus to the Sponge maven until it's tested).

Maven url for easy copying: `http://repo.mumfrey.com/content/repositories/mixin/`

**Comment by luacs1998** [comment]
Yeah I got it. I'm swamped in schoolwork so I'll only be able to test on Friday at the earliest.

Need to pair this with some work on #1470 anyway, as I mentioned in the issue on your repo.

Since I use the submodule method, there really was no need, I'll just manually check out the revision that was built on.

**Comment by Mumfrey** [comment]
Ah, I'd recommend switching to maven from now on since it'll make the shaded ASM dependency work a lot better (jarjar emits a somewhat broken jar which gets cleaned up in the final output build. And to be honest depending on mixin as a submodule is not really useful any more anyway.

**Comment by luacs1998** [comment]
Alright, noted. As a side, it would be nice to have a stable compile target ( -SNAPSHOT will keep changing to the latest build, which is rather undesirable to me...)

**Comment by Mumfrey** [comment]
That can certainly be arranged.

**Comment by luacs1998** [comment]
A version of FE build 564, updated to Mixin's shade-asm branch and some classloader changes, is available for testing here: http://www.mediafire.com/download/ft9s13nfrq2a3et/forgeessentials-1.7.10-1.4.0-beta7-derpy_mixins-server.jar

**Comment by BakermanLP** [comment]
Tested the mentioned build (...beta7-derpy...) and got the following errors on startup:
http://pastebin.com/nAmsrLdM

I also tried to remove fastcraft. After that I got a bunch of error messages at the end of startup. Here is one of them:
http://pastebin.com/zh8Jzbm7

Also in this version there seems to be the bug about "Saving permissions loop" ( ForgeEssentials/ForgeEssentialsMain#1491 )

**Comment by melekaiah** [comment]
also tested the derpy build and was unable to launch.

**Comment by luacs1998** [comment]
@BakermanLP try adding -Dfastcraft.asm.permissive=true to your command line logs. Your second error message seems to be about bad packets, I'll need you to make a new issue about that.

@melekaiah logs or it didn't happen.

**Comment by Ganjalf-Amsterdoom** [comment]
The afore mentioned build works, its spamming in console "saving permissions" and playerlogger crashed on first run, it loaded on a second load, running on a regrowth modpack with forge 1291.

**Comment by melekaiah** [comment]
Gonna download a fresh copy and test tonight when users are offline

**Comment by BLSK** [comment]
My Progression Mod issue seems to be working finally.  Seems FE beta7 crashes on first run...I reran the server and it works!...  I'll still be fiddling around making sure everything is up to snuff...but for a start its good

**Comment by olee** [comment]
@luacs1998 we should definitely investigate that first-run-crash and #1489 
Other than that it seems to work fine with any mod except fastcraft now.

**Comment by BLSK** [comment]
Is there any info you may need?  I would be helpful if we can get bugs worked out.  I am finding FE is a bit of a treasure box now that I am finding there is like a minor worldguard in it.

**Comment by melekaiah** [comment]
With fastcraft off it seems to be working prwtty gopd at the moment.

**Comment by ultrasn0w** [comment]
For me it's working fine even with fastcraft, just add -Dfastcraft.asm.permissve=true to your launch arguments. (This actually does not not turn fastcraft off)

**Comment by luacs1998** [comment]
Seems to be fixed, so I'll close this.

I know about fastcraft, the fix is to add -Dfastcraft.asm.permissive=true to your server args.
